File: android/app/tests/unit/src/com/android/bluetooth/hearingaid/HearingAidStateMachineTest.java
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.hearingaid;

import static android.bluetooth.BluetoothHearingAid.ACTION_CONNECTION_STATE_CHANGED;
import static android.bluetooth.BluetoothProfile.EXTRA_STATE;
import static android.bluetooth.BluetoothProfile.STATE_CONNECTED;
import static android.bluetooth.BluetoothProfile.STATE_CONNECTING;
import static android.bluetooth.BluetoothProfile.STATE_DISCONNECTED;

import static androidx.test.espresso.intent.matcher.IntentMatchers.hasAction;
import static androidx.test.espresso.intent.matcher.IntentMatchers.hasExtra;

import static com.android.bluetooth.TestUtils.MockitoRule;
import static com.android.bluetooth.TestUtils.getTestDevice;

import static com.google.common.truth.Truth.assertThat;

import static org.mockito.Mockito.any;
import static org.mockito.Mockito.anyString;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.eq;
import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

import android.bluetooth.BluetoothDevice;
import android.content.Intent;
import android.os.UserHandle;

import androidx.test.filters.SmallTest;
import androidx.test.runner.AndroidJUnit4;

import com.android.bluetooth.TestLooper;

import org.hamcrest.Matcher;
import org.hamcrest.core.AllOf;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.hamcrest.MockitoHamcrest;

/** Test cases for {@link HearingAidStateMachine}. */
@SmallTest
@RunWith(AndroidJUnit4.class)
public class HearingAidStateMachineTest {
    @Rule public final MockitoRule mMockitoRule = new MockitoRule();

    @Mock private HearingAidService mService;
    @Mock private HearingAidNativeInterface mNativeInterface;

    private final BluetoothDevice mTestDevice = getTestDevice(0xDA);

    private HearingAidStateMachine mStateMachine;
    private InOrder mInOrder;
    private TestLooper mLooper;

    @Before
    public void setUp() {
        mInOrder = inOrder(mService);
        mLooper = new TestLooper();

        doReturn(true).when(mService).okToConnect(any());
        doReturn(true).when(mService).isConnectedPeerDevices(mTestDevice);

        doReturn(true).when(mNativeInterface).connectHearingAid(any());
        doReturn(true).when(mNativeInterface).disconnectHearingAid(any());

        mStateMachine =
                new HearingAidStateMachine(
                        mService, mTestDevice, mNativeInterface, mLooper.getLooper());
        mStateMachine.start();
    }

    @Test
    public void initialState_isDisconnected() {
        assertThat(mStateMachine.getConnectionState()).isEqualTo(STATE_DISCONNECTED);
    }

    @Test
    public void incomingConnect_whenNotOkToConnect_isRejected() {
        doReturn(false).when(mService).okToConnect(any());

        // Inject an event for when incoming connection is requested
        HearingAidStackEvent connStCh =
                new HearingAidStackEvent(HearingAidStackEvent.EVENT_TYPE_CONNECTION_STATE_CHANGED);
        connStCh.device = mTestDevice;
        connStCh.valueInt1 = STATE_CONNECTED;
        sendAndDispatchMessage(HearingAidStateMachine.MESSAGE_STACK_EVENT, connStCh);

        verify(mService, never()).sendBroadcastAsUser(any(), any(), anyString(), any());
        assertThat(mStateMachine.getCurrentState())
                .isInstanceOf(HearingAidStateMachine.Disconnected.class);
    }

    @Test
    public void incomingConnect_whenOkToConnect_isConnected() {
        // Inject an event for when incoming connection is requested
        HearingAidStackEvent connStCh =
                new HearingAidStackEvent(HearingAidStackEvent.EVENT_TYPE_CONNECTION_STATE_CHANGED);
        connStCh.device = mTestDevice;
        connStCh.valueInt1 = STATE_CONNECTING;
        sendAndDispatchMessage(HearingAidStateMachine.MESSAGE_STACK_EVENT, connStCh);

        verifyIntentSent(
                hasAction(ACTION_CONNECTION_STATE_CHANGED),
                hasExtra(EXTRA_STATE, STATE_CONNECTING));

        assertThat(mStateMachine.getCurrentState())
                .isInstanceOf(HearingAidStateMachine.Connecting.class);

        // Send a message to trigger connection completed
        HearingAidStackEvent connCompletedEvent =
                new HearingAidStackEvent(HearingAidStackEvent.EVENT_TYPE_CONNECTION_STATE_CHANGED);
        connCompletedEvent.device = mTestDevice;
        connCompletedEvent.valueInt1 = STATE_CONNECTED;
        sendAndDispatchMessage(HearingAidStateMachine.MESSAGE_STACK_EVENT, connCompletedEvent);

        verifyIntentSent(
                hasAction(ACTION_CONNECTION_STATE_CHANGED), hasExtra(EXTRA_STATE, STATE_CONNECTED));
        assertThat(mStateMachine.getCurrentState())
                .isInstanceOf(HearingAidStateMachine.Connected.class);
    }

    @Test
    public void outgoingConnect_whenTimeOut_isDisconnectedAndInAcceptList() {
        sendAndDispatchMessage(HearingAidStateMachine.MESSAGE_CONNECT, mTestDevice);

        verifyIntentSent(
                hasAction(ACTION_CONNECTION_STATE_CHANGED),
                hasExtra(EXTRA_STATE, STATE_CONNECTING));
        assertThat(mStateMachine.getCurrentState())
                .isInstanceOf(HearingAidStateMachine.Connecting.class);

        mLooper.moveTimeForward(HearingAidStateMachine.CONNECT_TIMEOUT.toMillis());
        mLooper.dispatchAll();

        verifyIntentSent(
                hasAction(ACTION_CONNECTION_STATE_CHANGED),
                hasExtra(EXTRA_STATE, STATE_DISCONNECTED));
        assertThat(mStateMachine.getCurrentState())
                .isInstanceOf(HearingAidStateMachine.Disconnected.class);

        verify(mNativeInterface).addToAcceptlist(eq(mTestDevice));
    }

    @Test
    public void incomingConnect_whenTimeOut_isDisconnectedAndInAcceptList() {
        HearingAidStackEvent connStCh =
                new HearingAidStackEvent(HearingAidStackEvent.EVENT_TYPE_CONNECTION_STATE_CHANGED);
        connStCh.device = mTestDevice;
        connStCh.valueInt1 = STATE_CONNECTING;
        sendAndDispatchMessage(HearingAidStateMachine.MESSAGE_STACK_EVENT, connStCh);

        verifyIntentSent(
                hasAction(ACTION_CONNECTION_STATE_CHANGED),
                hasExtra(EXTRA_STATE, STATE_CONNECTING));
        assertThat(mStateMachine.getCurrentState())
                .isInstanceOf(HearingAidStateMachine.Connecting.class);

        mLooper.moveTimeForward(HearingAidStateMachine.CONNECT_TIMEOUT.toMillis());
        mLooper.dispatchAll();

        verifyIntentSent(
                hasAction(ACTION_CONNECTION_STATE_CHANGED),
                hasExtra(EXTRA_STATE, STATE_DISCONNECTED));
        assertThat(mStateMachine.getCurrentState())
                .isInstanceOf(HearingAidStateMachine.Disconnected.class);

        verify(mNativeInterface).addToAcceptlist(eq(mTestDevice));
    }

    private void sendAndDispatchMessage(int what, Object obj) {
        mStateMachine.sendMessage(what, obj);
        mLooper.dispatchAll();
    }

    @SafeVarargs
    private void verifyIntentSent(Matcher<Intent>... matchers) {
        mInOrder.verify(mService)
                .sendBroadcastAsUser(
                        MockitoHamcrest.argThat(AllOf.allOf(matchers)),
                        eq(UserHandle.ALL),
                        any(),
                        any());
    }
}

File: android/app/tests/unit/src/com/android/bluetooth/hearingaid/HearingAidStateMachineTest.kt
/*
 * Copyright (C) 2026 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.bluetooth.hearingaid

import android.bluetooth.BluetoothHearingAid.ACTION_CONNECTION_STATE_CHANGED
import android.bluetooth.BluetoothProfile.EXTRA_STATE
import android.bluetooth.BluetoothProfile.STATE_CONNECTED
import android.bluetooth.BluetoothProfile.STATE_CONNECTING
import android.bluetooth.BluetoothProfile.STATE_DISCONNECTED
import android.content.Intent
import android.os.Bundle
import androidx.test.espresso.intent.matcher.IntentMatchers.hasAction
import androidx.test.espresso.intent.matcher.IntentMatchers.hasExtra
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.SmallTest
import com.android.bluetooth.TestLooper
import com.android.bluetooth.getTestDevice
import com.android.bluetooth.hearingaid.HearingAidStateMachine.MESSAGE_CONNECTION_STATE_CHANGED
import com.android.tests.bluetooth.MockitoRule
import com.google.common.truth.Truth.assertThat
import org.hamcrest.Matcher
import org.hamcrest.core.AllOf
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.InOrder
import org.mockito.Mock
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.inOrder
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import org.mockito.hamcrest.MockitoHamcrest
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.whenever

/** Test cases for [HearingAidStateMachine]. */
@SmallTest
@RunWith(AndroidJUnit4::class)
class HearingAidStateMachineTest {
    @get:Rule val mockitoRule = MockitoRule()

    @Mock private lateinit var service: HearingAidService
    @Mock private lateinit var nativeInterface: HearingAidNativeInterface

    private val device = getTestDevice(0xDA)

    private lateinit var stateMachine: HearingAidStateMachine
    private lateinit var inOrder: InOrder
    private lateinit var looper: TestLooper

    @Before
    fun setUp() {
        inOrder = inOrder(service)
        looper = TestLooper()

        doReturn(true).whenever(service).okToConnect(any())
        doReturn(true).whenever(service).isConnectedPeerDevices(device)
        doReturn(true).whenever(nativeInterface).connectHearingAid(any())
        doReturn(true).whenever(nativeInterface).disconnectHearingAid(any())

        stateMachine = HearingAidStateMachine(service, device, nativeInterface, looper.looper)
        stateMachine.start()
    }

    @Test
    fun initialState_isDisconnected() {
        assertThat(stateMachine.connectionState).isEqualTo(STATE_DISCONNECTED)
    }

    @Test
    fun incomingConnect_whenNotOkToConnect_isRejected() {
        doReturn(false).whenever(service).okToConnect(any())
        stateMachine.sendMessage(MESSAGE_CONNECTION_STATE_CHANGED, STATE_CONNECTED)
        looper.dispatchAll()

        verify(service, never()).sendBroadcastAsUser(any(), any(), any<String>(), any<Bundle>())
        assertThat(stateMachine.currentState)
            .isInstanceOf(HearingAidStateMachine.Disconnected::class.java)
    }

    @Test
    fun incomingConnect_whenOkToConnect_isConnected() {
        stateMachine.sendMessage(MESSAGE_CONNECTION_STATE_CHANGED, STATE_CONNECTING)
        looper.dispatchAll()

        verifyIntentSent(
            hasAction(ACTION_CONNECTION_STATE_CHANGED),
            hasExtra(EXTRA_STATE, STATE_CONNECTING),
        )

        assertThat(stateMachine.currentState)
            .isInstanceOf(HearingAidStateMachine.Connecting::class.java)

        stateMachine.sendMessage(MESSAGE_CONNECTION_STATE_CHANGED, STATE_CONNECTED)
        looper.dispatchAll()

        verifyIntentSent(
            hasAction(ACTION_CONNECTION_STATE_CHANGED),
            hasExtra(EXTRA_STATE, STATE_CONNECTED),
        )
        assertThat(stateMachine.currentState)
            .isInstanceOf(HearingAidStateMachine.Connected::class.java)
    }

    @Test
    fun outgoingConnect_whenTimeOut_isDisconnectedAndInAcceptList() {
        sendAndDispatchMessage(HearingAidStateMachine.MESSAGE_CONNECT, device)

        verifyIntentSent(
            hasAction(ACTION_CONNECTION_STATE_CHANGED),
            hasExtra(EXTRA_STATE, STATE_CONNECTING),
        )
        assertThat(stateMachine.currentState)
            .isInstanceOf(HearingAidStateMachine.Connecting::class.java)

        looper.moveTimeForward(HearingAidStateMachine.CONNECT_TIMEOUT.toMillis())
        looper.dispatchAll()

        verifyIntentSent(
            hasAction(ACTION_CONNECTION_STATE_CHANGED),
            hasExtra(EXTRA_STATE, STATE_DISCONNECTED),
        )
        assertThat(stateMachine.currentState)
            .isInstanceOf(HearingAidStateMachine.Disconnected::class.java)

        verify(nativeInterface).addToAcceptlist(eq(device))
    }

    @Test
    fun incomingConnect_whenTimeOut_isDisconnectedAndInAcceptList() {
        stateMachine.sendMessage(MESSAGE_CONNECTION_STATE_CHANGED, STATE_CONNECTING)
        looper.dispatchAll()

        verifyIntentSent(
            hasAction(ACTION_CONNECTION_STATE_CHANGED),
            hasExtra(EXTRA_STATE, STATE_CONNECTING),
        )
        assertThat(stateMachine.currentState)
            .isInstanceOf(HearingAidStateMachine.Connecting::class.java)

        looper.moveTimeForward(HearingAidStateMachine.CONNECT_TIMEOUT.toMillis())
        looper.dispatchAll()

        verifyIntentSent(
            hasAction(ACTION_CONNECTION_STATE_CHANGED),
            hasExtra(EXTRA_STATE, STATE_DISCONNECTED),
        )
        assertThat(stateMachine.currentState)
            .isInstanceOf(HearingAidStateMachine.Disconnected::class.java)

        verify(nativeInterface).addToAcceptlist(eq(device))
    }

    @Test
    fun testDoQuit_inConnectedState_broadcastsDisconnectedIntent() {
        stateMachine.sendMessage(MESSAGE_CONNECTION_STATE_CHANGED, STATE_CONNECTING)
        looper.dispatchAll()
        stateMachine.sendMessage(MESSAGE_CONNECTION_STATE_CHANGED, STATE_CONNECTED)
        looper.dispatchAll()

        stateMachine.doQuit()

        verifyIntentSent(
            hasAction(ACTION_CONNECTION_STATE_CHANGED),
            hasExtra(EXTRA_STATE, STATE_DISCONNECTED),
        )
    }

    private fun sendAndDispatchMessage(what: Int, obj: Any?) {
        stateMachine.sendMessage(what, obj)
        looper.dispatchAll()
    }

    @SafeVarargs
    private fun verifyIntentSent(vararg matchers: Matcher<Intent?>?) {
        inOrder
            .verify(service)
            .sendBroadcast(
                MockitoHamcrest.argThat(AllOf.allOf(*matchers)),
                any<String>(),
                any<Bundle>(),
            )
    }
}
